sdk/rust/nym-sdk/src/stream_wrapper/mixnet_stream_wrapper.rs

```diff
@@ -0,0 +1,523 @@
+        let client = match socket {
+            Some(socket) => socket.into_inner(),
+            None => MixnetClient::connect_new().await.unwrap(),
+        };
+        Self {
+            client,
+            peer: Some(peer),
+            peer_surbs: None,
+        }
+    }
+
+    /// Nym address of Stream's peer (Nym Client it will communicate with).
+    pub fn peer_addr(&self) -> Recipient {
+        let peer = &self.peer.expect("No Peer set");
+        *peer
+    }
+
+    /// Our Nym address.
+    pub fn local_addr(&self) -> &Recipient {
+        self.client.nym_address()
+    }
+
+    /// Store SURBs in own field.
+    pub fn store_surbs(&mut self, surbs: AnonymousSenderTag) {
+        self.peer_surbs = Some(surbs);
+    }
+
+    /// Get SURBs (if any).
+    pub fn surbs(&self) -> Option<AnonymousSenderTag> {
+        self.peer_surbs
+    }
+
+    /// Split for concurrent read/write (like TcpStream::Split) into MixnetStreamReader and MixnetStreamWriter.
+    pub fn split(self) -> (MixStreamReader, MixStreamWriter) {
+        debug!("Splitting MixStream");
+        let sender = self.client.split_sender();
+        debug!("Split MixStream into Reader and Writer");
+        let (surb_tx, surb_rx) = oneshot::channel();
+        (
+            MixStreamReader {
+                client: self.client,
+                peer: self.peer,
+                peer_surbs: self.peer_surbs,
+                surb_tx: Some(surb_tx),
+            },
+            MixStreamWriter {
+                sender,
+                peer: self.peer.expect("No Peer set"),
+                peer_surbs: self.peer_surbs,
+                surb_rx: Some(surb_rx),
+            },
+        )
+    }
+
+    /// Convenience method for just piping bytes into the Mixnet.
+    pub async fn write_bytes(&mut self, data: &[u8]) -> Result<(), Error> {
+        let input_message = if self.peer_surbs.is_some() {
+            info!("Writing {} bytes, sending with SURBs", data.len());
+            InputMessage::Reply {
+                recipient_tag: (self.peer_surbs.expect("No Peer SURBs set")),
+                data: (data.to_owned()),
+                lane: (nym_task::connections::TransmissionLane::General),
+                max_retransmissions: (Some(5)), // TODO check with Drazen - guessing here
+            }
+        } else {
+            info!("Writing {} bytes", data.len());
+            InputMessage::Anonymous {
+                recipient: (self.peer.expect("No Peer set")),
+                data: (data.to_owned()),
+                reply_surbs: (10),
+                lane: (nym_task::connections::TransmissionLane::General),
+                max_retransmissions: (Some(5)), // TODO check with Drazen - guessing here
+            }
+        };
+
+        let mut codec = InputMessageCodec {};
+        let mut serialized_bytes = BytesMut::new();
+        codec.encode(input_message, &mut serialized_bytes)?;
+        debug!("Serialized bytes: {:?}", serialized_bytes);
```

Comment:
providing actual bytes in the debug output could be quite annoying

Reply:
I could either switch to trace or just remove it, that's true